src/main/java/org/opensearch/flowframework/workflow/WorkflowStepFactory.java

```diff
@@ -42,21 +43,21 @@ public WorkflowStepFactory(
         MachineLearningNodeClient mlClient,
         FlowFrameworkIndicesHandler flowFrameworkIndicesHandler
     ) {
-        stepMap.put(NoOpStep.NAME, new NoOpStep());
-        stepMap.put(CreateIndexStep.NAME, new CreateIndexStep(clusterService, client, flowFrameworkIndicesHandler));
-        stepMap.put(CreateIngestPipelineStep.NAME, new CreateIngestPipelineStep(client, flowFrameworkIndicesHandler));
+        stepMap.put(NoOpStep.NAME, NoOpStep::new);
+        stepMap.put(CreateIndexStep.NAME, () -> new CreateIndexStep(clusterService, client, flowFrameworkIndicesHandler));
+        stepMap.put(CreateIngestPipelineStep.NAME, () -> new CreateIngestPipelineStep(client, flowFrameworkIndicesHandler));
         stepMap.put(
             RegisterLocalModelStep.NAME,
-            new RegisterLocalModelStep(settings, clusterService, mlClient, flowFrameworkIndicesHandler)
+            () -> new RegisterLocalModelStep(settings, clusterService, mlClient, flowFrameworkIndicesHandler)
         );
-        stepMap.put(RegisterRemoteModelStep.NAME, new RegisterRemoteModelStep(mlClient, flowFrameworkIndicesHandler));
-        stepMap.put(DeployModelStep.NAME, new DeployModelStep(mlClient));
-        stepMap.put(UndeployModelStep.NAME, new UndeployModelStep(mlClient));
-        stepMap.put(CreateConnectorStep.NAME, new CreateConnectorStep(mlClient, flowFrameworkIndicesHandler));
-        stepMap.put(DeleteConnectorStep.NAME, new DeleteConnectorStep(mlClient));
-        stepMap.put(ModelGroupStep.NAME, new ModelGroupStep(mlClient, flowFrameworkIndicesHandler));
-        stepMap.put(ToolStep.NAME, new ToolStep());
-        stepMap.put(RegisterAgentStep.NAME, new RegisterAgentStep(mlClient));
+        stepMap.put(RegisterRemoteModelStep.NAME, () -> new RegisterRemoteModelStep(mlClient, flowFrameworkIndicesHandler));
+        stepMap.put(DeployModelStep.NAME, () -> new DeployModelStep(mlClient));
+        stepMap.put(UndeployModelStep.NAME, () -> new UndeployModelStep(mlClient));
+        stepMap.put(CreateConnectorStep.NAME, () -> new CreateConnectorStep(mlClient, flowFrameworkIndicesHandler));
+        stepMap.put(DeleteConnectorStep.NAME, () -> new DeleteConnectorStep(mlClient));
+        stepMap.put(ModelGroupStep.NAME, () -> new ModelGroupStep(mlClient, flowFrameworkIndicesHandler));
+        stepMap.put(ToolStep.NAME, ToolStep::new);
+        stepMap.put(RegisterAgentStep.NAME, () -> new RegisterAgentStep(mlClient));
```

Comment:
We should refer the method here as well. There can be multiple agents.

Reply:
Ignore this comment.

Reply:
> Ignore this comment.

No.